anadama/tm/configuration_parameters.txt
{
    "TEMP_PATH": "/tmp/batch_output",
    "SOURCE_PATH": "/aux/deploy6/bin/activate"
    "GOVERNOR" : 10,
    "PORT" : 8888,
    "HOOKS" : "/aux/deploy6/tmp/ccfa/mibc/tm/hooks",
    "CLUSTER": "LSF", 
    "CLUSTER_OUTPUT_PARAM": "-o", 
    "CLUSTER_JOB": "bsub",
    "CLUSTER_PROJECT": " -P unspecified--broadfolk",
    "CLUSTER_MEMORY": [ 
        " -R rusage[mem=4]",
        " -R rusage[mem=8]" ],
    "CLUSTER_QUEUE": [
        " -q bhour",
        " -q bweek" ],
    "CLUSTER_JOBNAME": " -J ",
    "CLUSTER_QUERY": "bjobs -noheader -w", 
    "CLUSTER_STATS": "bjobs -l ",
    "CLUSTER_JOBID_POS": 2,
    "CLUSTER_STATUS_POS": 3
}


anadama/tm/configuration_supervisord.txt
; Sample supervisor config file.
;
; For more information on the config file, please see:
; http://supervisord.org/configuration.html
;
; Notes:
;  - Shell expansion ("~" or "$HOME") is not supported.  Environment
;    variables can be expanded using this syntax: "%(ENV_HOME)s".
;  - Comments must have a leading space: "a=b ;comment" not "a=b;comment".

[unix_http_server]
file=/tmp/supervisor.sock   ; (the path to the socket file)
;chmod=0700                 ; socket file mode (default 0700)
;chown=nobody:nogroup       ; socket file uid:gid owner
;username=user              ; (default is no username (open server))
;password=123               ; (default is no password (open server))

;[inet_http_server]         ; inet (TCP) server disabled by default
;port=127.0.0.1:9001        ; (ip_address:port specifier, *:port for all iface)
;username=user              ; (default is no username (open server))
;password=123               ; (default is no password (open server))

[supervisord]
;logfile=/tmp/supervisord.log ; (main log file;default $CWD/supervisord.log)
logfile_maxbytes=5MB         ; (max main logfile bytes b4 rotation;default 50MB)
logfile_backups=2            ; (num of main logfile rotation backups;default 10)
loglevel=info                ; (log level;default info; others: debug,warn,trace)
pidfile=supervisord.pid      ; (supervisord pidfile;default supervisord.pid)
nodaemon=false               ; (start in foreground if true;default false)
minfds=1024                  ; (min. avail startup file descriptors;default 1024)
minprocs=200                 ; (min. avail process descriptors;default 200)
;umask=022                   ; (process file creation umask;default 022)
;user=chrism                 ; (default is current user, required if root)
;identifier=supervisor       ; (supervisord identifier, default is 'supervisor')
;directory=/tmp              ; (default is not to cd during start)
;nocleanup=true              ; (don't clean up tempfiles at start;default false)
;childlogdir=/tmp            ; ('AUTO' child log dir, default $TEMP)
;environment=KEY="value"     ; (key value pairs to add to environment)
;strip_ansi=false            ; (strip ansi escape codes in logs; def. false)

; the below section must remain in the config file for RPC
; (supervisorctl/web interface) to work, additional interfaces may be
; added by defining them in separate rpcinterface: sections
;[rpcinterface:supervisor]
;supervisor.rpcinterface_factory = supervisor.rpcinterface:make_main_rpcinterface

[supervisorctl]
serverurl=unix:///$CWD/supervisor.sock ; use a unix:// URL  for a unix socket
;serverurl=http://127.0.0.1:9001 ; use an http:// url to specify an inet socket
;username=chris              ; should be same as http_username if set
;password=123                ; should be same as http_password if set
;prompt=mysupervisor         ; cmd line prompt (default "supervisor")
;history_file=~/.sc_history  ; use readline history if available

; The below sample program section shows all possible program subsection values,
; create one or more 'real' program: sections to be able to control them under
; supervisor.

[program:mibc_tm_daemon]
command=mibc_tm_daemon    ; the program (relative uses PATH, can take args)
;process_name=%(program_name)s ; process_name expr (default %(program_name)s)
numprocs=1                    ; number of processes copies to start (def 1)
;directory=/tmp                ; directory to cwd to before exec (def no cwd)
;umask=022                     ; umask for process (default None)
;priority=999                  ; the relative start priority (default 999)
autostart=true                ; start at supervisord start (default: true)
autorestart=unexpected        ; whether/when to restart (default: unexpected)
;startsecs=1                   ; number of secs prog must stay running (def. 1)
;startretries=3                ; max # of serial start failures (default 3)
;exitcodes=0,2                 ; 'expected' exit codes for process (default 0,2)
;stopsignal=QUIT               ; signal used to kill process (default TERM)
;stopwaitsecs=10               ; max num secs to wait b4 SIGKILL (default 10)
;stopasgroup=false             ; send stop signal to the UNIX process group (default false)
;killasgroup=false             ; SIGKILL the UNIX process group (def false)
;user=chrism                   ; setuid to this UNIX account to run the program
redirect_stderr=true          ; redirect proc stderr to stdout (default false)
;stdout_logfile=/a/path        ; stdout log path, NONE for none; default AUTO
;stdout_logfile_maxbytes=1MB   ; max # logfile bytes b4 rotation (default 50MB)
;stdout_logfile_backups=10     ; # of stdout logfile backups (default 10)
;stdout_capture_maxbytes=1MB   ; number of bytes in 'capturemode' (default 0)
;stdout_events_enabled=false   ; emit events on stdout writes (default false)
;stderr_logfile=/a/path        ; stderr log path, NONE for none; default AUTO
;stderr_logfile_maxbytes=1MB   ; max # logfile bytes b4 rotation (default 50MB)
;stderr_logfile_backups=10     ; # of stderr logfile backups (default 10)
;stderr_capture_maxbytes=1MB   ; number of bytes in 'capturemode' (default 0)
;stderr_events_enabled=false   ; emit events on stderr writes (default false)
;environment=A="1",B="2"       ; process environment additions (def no adds)
;serverurl=AUTO                ; override serverurl computation (childutils)

; The below sample eventlistener section shows all possible
; eventlistener subsection values, create one or more 'real'
; eventlistener: sections to be able to handle event notifications
; sent by supervisor.

;[eventlistener:theeventlistenername]
;command=/bin/eventlistener    ; the program (relative uses PATH, can take args)
;process_name=%(program_name)s ; process_name expr (default %(program_name)s)
;numprocs=1                    ; number of processes copies to start (def 1)
;events=EVENT                  ; event notif. types to subscribe to (req'd)
;buffer_size=10                ; event buffer queue size (default 10)
;directory=/tmp                ; directory to cwd to before exec (def no cwd)
;umask=022                     ; umask for process (default None)
;priority=-1                   ; the relative start priority (default -1)
;autostart=true                ; start at supervisord start (default: true)
;autorestart=unexpected        ; whether/when to restart (default: unexpected)
;startsecs=1                   ; number of secs prog must stay running (def. 1)
;startretries=3                ; max # of serial start failures (default 3)
;exitcodes=0,2                 ; 'expected' exit codes for process (default 0,2)
;stopsignal=QUIT               ; signal used to kill process (default TERM)
;stopwaitsecs=10               ; max num secs to wait b4 SIGKILL (default 10)
;stopasgroup=false             ; send stop signal to the UNIX process group (default false)
;killasgroup=false             ; SIGKILL the UNIX process group (def false)
;user=chrism                   ; setuid to this UNIX account to run the program
;redirect_stderr=true          ; redirect proc stderr to stdout (default false)
;stdout_logfile=/a/path        ; stdout log path, NONE for none; default AUTO
;stdout_logfile_maxbytes=1MB   ; max # logfile bytes b4 rotation (default 50MB)
;stdout_logfile_backups=10     ; # of stdout logfile backups (default 10)
;stdout_events_enabled=false   ; emit events on stdout writes (default false)
;stderr_logfile=/a/path        ; stderr log path, NONE for none; default AUTO
;stderr_logfile_maxbytes=1MB   ; max # logfile bytes b4 rotation (default 50MB)
;stderr_logfile_backups        ; # of stderr logfile backups (default 10)
;stderr_events_enabled=false   ; emit events on stderr writes (default false)
;environment=A="1",B="2"       ; process environment additions
;serverurl=AUTO                ; override serverurl computation (childutils)

; The below sample group section shows all possible group values,
; create one or more 'real' group: sections to create "heterogeneous"
; process groups.

;[group:thegroupname]
;programs=progname1,progname2  ; each refers to 'x' in [program:x] definitions
;priority=999                  ; the relative start priority (default 999)

; The [include] section can just contain the "files" setting.  This
; setting can list multiple files (separated by whitespace or
; newlines).  It can also contain wildcards.  The filenames are
; interpreted as relative to this file.  Included files *cannot*
; include files themselves.

;[include]
;files = relative/directory/*.ini


anadama/tm/cli.py
#!/usr/bin/env python
import optparse, sys, json, pprint
import tasks, parser
import tm as TM
import tm_daemon
import os
import signal
import tempfile
from pprint import pprint
import websocket
import re
import hashlib
import globals


HELP="""%prog -t type [-i <json encoded inputfile>] [-l location] [-D] [-g governor] [-p port][-k hooks dir]

%prog - TM (Task Manager) parses the tasks contained in the given 
json encoded directed acyclic graph (DAG) file.  This command
is designed to read the json encoded DAG from stdin.  However, a filename 
may be specified for input (-i filename.json) if that method is preferred.
type currently can be one of three values: local, slurm, or lsf.

User supplied 'hooks' are called for the given events:
    post_pipeline_failure.sh  
    post_pipeline_success.sh  
    post_task_failure.sh  
    post_task_success.sh  
    pre_pipeline.sh  
    pre_pipeline_restart.sh
    

NOTE:
-D indicates that the task_manager daemon should be started and will 
process all DAGs sent via a websocket interface to the daemon listening 
on the given '-p' port.  If no daemon is listening on that port, a new
daemon will be started.
"""

opts_list = [
    optparse.make_option('-v', '--verbose', action="store_true",
                         dest="verbose", default=False,
                         help="Turn on verbose output (to stderr)"),
    optparse.make_option('-i', '--input', action="store", type="string",
                         dest="dagfile", help="Specify json encrypted dag file"),
    optparse.make_option('-l', '--location', action="store", type="string", default="",
                         dest="location", help="Specify directory where all tasks and logfiles are stored.  Default is the current working directory."),
    optparse.make_option('-t', '--type', action="store",
                         dest="type", type="string", 
                         help="The type of task to be run (local, slurm, or lsf)"),
    optparse.make_option('-g', '--governor', action="store", type="int", 
                         dest="governor", help="Rate limit the number of concurrent tasks.  Useful for limited resource on local desktops / laptops."),
    optparse.make_option('-D', '--daemon', action="store_true", 
                         dest="daemon", help="Specify task manager to run as a daemon.  Will process multiple dags."),
    optparse.make_option('-p', '--port', action="store", type="int", 
                         dest="port", help="Specify the port of the webserver - defaults to 8888."),
    optparse.make_option('-k', '--hooks', action="store", type="string", 
                         dest="hooks", help="Specify the directory of the user supplied hooks.")
]


global opts, p, data, wslisteners
types = ('local', 'slurm', 'lsf')
wslisteners = []

def fileHandling():
   
    '''create directory structure to store metadata from tasks'''

    global hashdirectory, rundirectory

    hash, product = peekForHash(data)

    hashdirectory = opts.location + "/" + hash

    # create task directory if it doesn't exist
    run = "/run1"
    if not os.path.exists(hashdirectory):
        os.makedirs(hashdirectory)
        rundirectory = hashdirectory + run
    else:
        # get the last run and increment it by one
        runnum = 1
        for run in os.walk(hashdirectory).next()[1]:
            #print "run: " + run
            curnum = int(run[3:])
            if runnum < curnum:
                runnum = curnum
        runnum += 1;
        #print "runnum: " + str(runnum)
        rundirectory = hashdirectory + "/run" + str(runnum)

    os.makedirs(rundirectory)

    # create link to directory if it doesn't exist
    symlinkdir = opts.location + "/by_task_name"
    if not os.path.exists(symlinkdir):
        os.makedirs(symlinkdir)
        
    symlink = symlinkdir + "/" + product
    #print "symlink: " + symlink
    if not os.path.exists(symlink):
        print "symlink: " + hashdirectory + "  " + symlink
        os.symlink(hashdirectory, symlink)

def optionHandling():
    global opts, data

    if opts.dagfile:
        if not os.path.isfile(opts.dagfile):
            print >> sys.stderr, "opts.dagfile not found."
            argParser.print_usage()
            sys.exit(1)
        else:
            input = opts.dagfile
            if opts.verbose:
                print >> sys.stderr, "Input is: ", opts.dagfile
    else:
        input = "-" 
        if opts.verbose:
            print >> sys.stderr, "Input is stdin."

    if opts.type not in types:
        if globals.config.get("CLUSTER") is None:
            print >> sys.stderr, "type not one of 'local', 'slurm', or 'lsf'."
            argParser.print_usage()
            sys.exit(1)
        else:
            opts.type = globals.config["CLUSTER"].lower()

    if opts.location is "":
        if globals.config.get("TEMP_PATH") is not None:
            opts.location = globals.config["TEMP_PATH"] + "/anadama_flows"
        else:
            opts.location = os.getcwd() + "/anadama_flows"

    if not os.access(os.path.dirname(opts.location), os.W_OK):
        print >> sys.stderr, "directory " + opts.location + " is not writable."
        argParser.print_usage()
        sys.exit(1)

    if opts.port is None or "":
        if globals.config.get("PORT") is None:
            opts.port = 8888
        else:
            opts.port = globals.config["PORT"]

    if opts.governor is None or "":
        if globals.config.get("GOVERNOR") is None:
            opts.governor = 999
        else:
            opts.governor = globals.config["GOVERNOR"]

    if opts.hooks is None or "":
        if globals.config.get("HOOKS") is None:
            opts.hooks = os.path.dirname(os.path.realpath(__file__)) + "/hooks"
        else:
            opts.hooks = globals.config["HOOKS"]
            
    if input is "-":
        jsonString = sys.stdin.read()
        data = json.loads(jsonString);
    else:
        jsonfile = open(opts.dagfile)
        jsondata = jsonfile.read()
        data = json.loads(jsondata);
        jsonfile.close()

def peekForHash(data):
    nodes = data['dag']
    for entry in nodes:
        node = entry['node']
        if 'produces' in node and len(node['produces']) > 0:
            md5 = hashlib.md5()
            md5.update(node['produces'][0])
            name = node['name'].split(':')[0]
            return (str(md5.hexdigest())[:5], name)

                
def main():
    global opts, p, argParser, tm, config

    def sigtermSetup():
        signal.signal(signal.SIGTERM, sigtermHandler)
        signal.signal(signal.SIGINT, sigtermHandler)

    def sigtermHandler(signum, frame):
        print "caught signal " + str(signum)
        print "cleaning up..."
        tm_daemon.Tm_daemon.cleanup()
        print "shutting down webserver..."
        sys.exit(0)

    # load configuration parameters (if they exist)
    globals.init(os.getcwd())

    # read in options
    argParser = optparse.OptionParser(option_list=opts_list,
                                   usage=HELP)
    (opts, args) = argParser.parse_args()
    optionHandling()

    # setup output directories
    fileHandling()
   
    # get our installed location
    currentFile = os.path.realpath(__file__)
    path = os.path.dirname(currentFile)
    web_install_path = os.path.join(path, "anadama_flows")

    # create json to send to server
    d = {'tm': 
            {'dag': data, 
                'type': opts.type, 
                'location': opts.location,
                'rundirectory': rundirectory,
                'hashdirectory': hashdirectory,
                'governor': opts.governor,
                'hooks': opts.hooks}}
    msg = json.dumps(d)
    #print >> sys.stderr, msg

    # if opts.daemon is false; call daemon directly to process single dag in foreground.
    if not opts.daemon:

        # first setup up signals
        sigtermSetup()

        daemon = tm_daemon.Tm_daemon()
        daemon.setupTm(d['tm'])
        daemon.run(opts.port)
        # daemon doesn't return

    # connect to daemon (or start it)
    else:
        done = False
        while not done:
            try:
                ws = websocket.create_connection("ws://localhost:" + str(opts.port) + "/websocket/")
                down = True
            except:
                print >> sys.stderr, "starting daemon..."
                subprocess.popen('supervisord', '-c', 'configuration_supervisord.conf')
                sleep(2)

        ws.send(msg)
        ws.close()

if __name__ == '__main__':
    main()
